File: api/gdflix.py
# /api/gdflix.py
from http.server import BaseHTTPRequestHandler
import json
from urllib.parse import urljoin, urlparse, parse_qs
import time
import re
import cloudscraper # Use cloudscraper instead of requests
from bs4 import BeautifulSoup
# ---- ADD THIS LINE ----
from requests.exceptions import Timeout, RequestException # Import specific exceptions

# --- Constants ---
GENERATION_TIMEOUT = 40
POLL_INTERVAL = 5
DEFAULT_HEADERS = {
    'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36',
    'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9',
    'Accept-Language': 'en-US,en;q=0.9',
}

# --- GDFLIX Bypass Logic (adapted to use cloudscraper) ---
def get_gdflix_download_link_with_scraper(scraper, start_url):
    """
    Fetches the final download link using a cloudscraper instance.
    Handles 'Generate' button by mimicking its POST request.
    """
    print(f"GDFLIX Bypass: Starting for URL: {start_url}")
    page1_url = start_url
    try:
        # --- Step 1 & 2 --- (Fetch and parse page 1)
        print(f"GDFLIX Bypass: Fetching initial URL: {page1_url}")
        response1 = scraper.get(page1_url, allow_redirects=True, timeout=30)
        response1.raise_for_status()
        page1_url = response1.url
        print(f"GDFLIX Bypass: Redirected to/Landed on: {page1_url}")
        soup1 = BeautifulSoup(response1.text, 'lxml')
        possible_tags_p1 = soup1.find_all(['a', 'button'])

        # --- Step 3 --- (Find Fast Cloud)
        fast_cloud_link_tag = None
        fast_cloud_pattern = re.compile(r'fast\s+cloud\s+download', re.IGNORECASE)
        for tag in possible_tags_p1:
            if fast_cloud_pattern.search(tag.get_text(strip=True)):
                fast_cloud_link_tag = tag
                print(f"GDFLIX Bypass: Found primary target (Fast Cloud): {tag.name}...")
                break

        # --- If Fast Cloud Found ---
        if fast_cloud_link_tag:
            # --- Steps 3a, 4, 5 --- (Getting to page 2)
            fast_cloud_href = fast_cloud_link_tag.get('href')
            if not fast_cloud_href and fast_cloud_link_tag.name == 'button':
                parent_form = fast_cloud_link_tag.find_parent('form')
                if parent_form: fast_cloud_href = parent_form.get('action')
            if not fast_cloud_href:
                print("GDFLIX Bypass Error: Found 'Fast Cloud Download' but couldn't get URL.")
                return None, "Fast Cloud element found, but no link associated."

            second_page_url = urljoin(page1_url, fast_cloud_href)
            print(f"GDFLIX Bypass: Found Fast Cloud link URL: {second_page_url}")
            time.sleep(1)

            print(f"GDFLIX Bypass: Fetching second page URL (Generate button page): {second_page_url}")
            fetch_headers_p2 = {'Referer': page1_url}
            response2 = scraper.get(second_page_url, timeout=30, headers=fetch_headers_p2)
            response2.raise_for_status()
            page2_url = response2.url
            print(f"GDFLIX Bypass: Landed on second page: {page2_url}")
            soup2 = BeautifulSoup(response2.text, 'lxml')
            possible_tags_p2 = soup2.find_all(['a', 'button'])

            # --- Step 6 --- (Find Cloud Resume Download)
            resume_link_tag = None
            resume_text_pattern = re.compile(r'cloud\s+resume\s+download', re.IGNORECASE)
            for tag in possible_tags_p2:
                 if resume_text_pattern.search(tag.get_text(strip=True)):
                    resume_link_tag = tag
                    print(f"GDFLIX Bypass: Found final link tag directly (Cloud Resume): {tag.name}...")
                    break

            # --- Step 6a --- (If found directly)
            if resume_link_tag:
                final_link_href = resume_link_tag.get('href')
                if not final_link_href and resume_link_tag.name == 'button':
                     parent_form = resume_link_tag.find_parent('form')
                     if parent_form: final_link_href = parent_form.get('action')
                if not final_link_href:
                    print("GDFLIX Bypass Error: Found 'Cloud Resume' but no href/action.")
                    return None, "Cloud Resume element found, but no link."
                final_download_link = urljoin(page2_url, final_link_href)
                print(f"GDFLIX Bypass: Found final Cloud Resume link URL: {final_download_link}")
                return final_download_link, None

            # --- Step 6b --- (If not found directly, check for Generate button by ID/Text)
            else:
                print("GDFLIX Bypass Info: 'Cloud Resume Download' not found directly. Checking for 'Generate Cloud Link' button...")
                generate_tag = soup2.find('button', id='cloud')
                if not generate_tag:
                    generate_pattern = re.compile(r'generate\s+cloud\s+link', re.IGNORECASE)
                    for tag in possible_tags_p2:
                        if generate_pattern.search(tag.get_text(strip=True)):
                            generate_tag = tag
                            break

                # --- If Generate button is found, MIMIC THE POST REQUEST ---
                if generate_tag:
                    print(f"GDFLIX Bypass: Found 'Generate Cloud Link' button: {generate_tag.name} id='{generate_tag.get('id', 'N/A')}'")
                    print("GDFLIX Bypass Info: Attempting to mimic the JavaScript POST request...")

                    post_data = {'action': 'cloud', 'key': '08df4425e31c4330a1a0a3cefc45c19e84d0a192', 'action_token': ''}
                    parsed_uri = urlparse(page2_url)
                    hostname = parsed_uri.netloc
                    post_headers = {'x-token': hostname, 'Referer': page2_url}

                    print(f"GDFLIX Bypass Info: Sending POST request to {page2_url}...")
                    page3_url = None
                    try:
                        post_response = scraper.post(page2_url, data=post_data, headers=post_headers, timeout=30)
                        post_response.raise_for_status()

                        try:
                            response_data = post_response.json()
                            print(f"GDFLIX Bypass Info: POST response JSON: {response_data}")
                            page3_url_relative = response_data.get('visit_url') or response_data.get('url')
                            if page3_url_relative:
                                page3_url = urljoin(page2_url, page3_url_relative)
                                print(f"GDFLIX Bypass Info: POST successful. Need to poll new URL: {page3_url}")
                            elif response_data.get('error'):
                                error_msg = response_data.get('message', 'Unknown POST error')
                                print(f"GDFLIX Bypass Error from POST request: {error_msg}")
                                return None, f"POST Error: {error_msg}"
                            else:
                                print("GDFLIX Bypass Error: POST response JSON format unknown.")
                                return None, "Unknown POST response format."
                        except json.JSONDecodeError:
                             if 300 <= post_response.status_code < 400 and 'location' in post_response.headers:
                                page3_url = urljoin(page2_url, post_response.headers['location'])
                                print(f"GDFLIX Bypass Info: POST resulted in redirect to: {page3_url}. Polling this.")
                             else:
                                print(f"GDFLIX Bypass Error: Failed to decode JSON response from POST. Status: {post_response.status_code}")
                                print("Response text (first 500 chars):", post_response.text[:500])
                                if "cloudflare" in post_response.text.lower() or "captcha" in post_response.text.lower():
                                    print("GDFLIX Bypass Hint: Cloudflare/Captcha challenge likely blocked the POST request.")
                                    return None, "Cloudflare/Captcha Blocked POST"
                                else:
                                    return None, f"Non-JSON POST response (Status: {post_response.status_code})"

                    except cloudscraper.exceptions.CloudflareException as cf_err:
                         print(f"GDFLIX Bypass Error: Cloudflare challenge during POST: {cf_err}")
                         return None, "Cloudflare Blocked POST"
                    # Use imported RequestException here
                    except RequestException as post_err:
                         print(f"GDFLIX Bypass Error during POST request: {post_err}")
                         return None, f"POST Network Error: {post_err}"

                    # --- If POST was successful and we have page3_url, START POLLING ---
                    if page3_url:
                        print(f"GDFLIX Bypass Info: Starting polling loop for {page3_url}...")
                        start_time = time.time()
                        while time.time() - start_time < GENERATION_TIMEOUT:
                            elapsed_time = time.time() - start_time
                            remaining_time = GENERATION_TIMEOUT - elapsed_time
                            if remaining_time <= 0: break

                            wait_time = min(POLL_INTERVAL, remaining_time)
                            print(f"GDFLIX Bypass Info: Waiting {wait_time:.1f}s before checking {page3_url}...")
                            time.sleep(wait_time)

                            try:
                                poll_headers = {'Referer': page3_url}
                                poll_response = scraper.get(page3_url, timeout=30, headers=poll_headers)
                                if poll_response.status_code != 200:
                                    print(f"GDFLIX Bypass Warning: Polling status {poll_response.status_code}. Retrying.")
                                    continue

                                poll_soup = BeautifulSoup(poll_response.text, 'lxml')
                                polled_resume_tag = None
                                for tag in poll_soup.find_all(['a', 'button']):
                                    if resume_text_pattern.search(tag.get_text(strip=True)):
                                        polled_resume_tag = tag
                                        print(f"\nGDFLIX Bypass Success: Found 'Cloud Resume Download' after polling!")
                                        break

                                if polled_resume_tag:
                                    final_link_href = polled_resume_tag.get('href')
                                    if not final_link_href and polled_resume_tag.name == 'button':
                                        parent_form = polled_resume_tag.find_parent('form')
                                        if parent_form: final_link_href = parent_form.get('action')
                                    if not final_link_href:
                                        print("GDFLIX Bypass Error: Found polled 'Cloud Resume' but no href/action.")
                                        return None, "Polled Cloud Resume found, but no link."
                                    final_download_link = urljoin(page3_url, final_link_href)
                                    print(f"GDFLIX Bypass: Found final Cloud Resume link URL after polling: {final_download_link}")
                                    return final_download_link, None

                            # Use imported RequestException here
                            except RequestException as poll_err:
                                print(f"GDFLIX Bypass Warning: Error during polling request: {poll_err}. Will retry.")
                            except Exception as parse_err:
                                print(f"GDFLIX Bypass Warning: Error parsing polled page: {parse_err}. Will retry.")

                        print(f"GDFLIX Bypass Error: Link generation timed out after {GENERATION_TIMEOUT}s.")
                        return None, f"Link is generating. Try again after a few minutes. (Timeout: {GENERATION_TIMEOUT}s)"

                else:
                    print("GDFLIX Bypass Error: Neither 'Cloud Resume Download' nor 'Generate Cloud Link' button found on the second page.")
                    return None, "Required button not found on intermediate page."

        # --- Step 3b: Fallback - PixeldrainDL on page 1 ---
        else:
            print("GDFLIX Bypass Info: 'Fast Cloud Download' not found. Checking for 'PixeldrainDL'...")
            pixeldrain_link_tag = None
            pixeldrain_pattern = re.compile(r'pixeldrain\s*dl', re.IGNORECASE)
            for tag in possible_tags_p1:
                if pixeldrain_pattern.search(tag.get_text(strip=True)):
                    pixeldrain_link_tag = tag
                    print(f"GDFLIX Bypass: Found fallback tag (Pixeldrain): {tag.name}...")
                    break

            if pixeldrain_link_tag:
                pixeldrain_href = pixeldrain_link_tag.get('href')
                if not pixeldrain_href and pixeldrain_link_tag.name == 'button':
                    parent_form = pixeldrain_link_tag.find_parent('form')
                    if parent_form: pixeldrain_href = parent_form.get('action')

                if pixeldrain_href:
                    pixeldrain_full_url = urljoin(page1_url, pixeldrain_href)
                    print(f"GDFLIX Bypass: Found Pixeldrain link URL: {pixeldrain_full_url}")
                    return pixeldrain_full_url, None
                else:
                    print("GDFLIX Bypass Error: Found Pixeldrain element but couldn't get href/action.")
                    return None, "Pixeldrain found, but no link."

            print("GDFLIX Bypass Error: Neither 'Fast Cloud Download' nor 'PixeldrainDL' link found/processed on the first page.")
            return None, "Primary and fallback links not found."


    except cloudscraper.exceptions.CloudflareException as cf_err:
        print(f"GDFLIX Bypass Error: Cloudflare challenge encountered: {cf_err}")
        return None, "Cloudflare protection blocked access."
    # Use imported Timeout and RequestException here
    except Timeout:
        print("GDFLIX Bypass Error: Request timed out.")
        return None, "Request timed out."
    except RequestException as e:
        print(f"GDFLIX Bypass Error: Network error during requests: {e}")
        return None, f"Network Error: {e}"
    except Exception as e:
        import traceback
        print(f"GDFLIX Bypass Error: An unexpected error occurred: {e}")
        traceback.print_exc()
        return None, f"Unexpected Error: {e}"

# --- API Handler Class (remains the same) ---
class handler(BaseHTTPRequestHandler):

    def do_POST(self):
        content_length = int(self.headers.get('Content-Length', 0))
        if content_length == 0:
            self._send_response(400, {"success": False, "error": "Missing request body"})
            return

        try:
            body = self.rfile.read(content_length)
            data = json.loads(body)
        except json.JSONDecodeError:
            self._send_response(400, {"success": False, "error": "Invalid JSON format"})
            return
        except Exception as e:
             self._send_response(400, {"success": False, "error": f"Error reading request body: {e}"})
             return

        gdflix_url = data.get('gdflixUrl')
        if not gdflix_url:
            self._send_response(400, {"success": False, "error": "Missing 'gdflixUrl' in request body"})
            return

        scraper = None
        try:
            print("GDFLIX API: Creating cloudscraper instance...")
            scraper = cloudscraper.create_scraper(
                 browser={
                    'browser': 'chrome',
                    'platform': 'windows',
                    'mobile': False
                }
            )
            scraper.headers.update(DEFAULT_HEADERS)
            print("GDFLIX API: Starting bypass process...")
            final_url, error_message = get_gdflix_download_link_with_scraper(scraper, gdflix_url)
            print("GDFLIX API: Bypass process finished.")

            if final_url:
                self._send_response(200, {"success": True, "finalUrl": final_url})
            else:
                status_code = 500
                if "Timeout" in (error_message or "") or "generating" in (error_message or ""):
                    status_code = 504
                elif "Cloudflare" in (error_message or "") or "Captcha" in (error_message or ""):
                     status_code = 503
                elif "not found" in (error_message or "").lower() or "no link" in (error_message or "").lower():
                     status_code = 404

                self._send_response(status_code, {"success": False, "error": error_message or "Bypass failed for unknown reason."})

        except Exception as e:
            import traceback
            print("GDFLIX API: Unhandled exception during bypass!")
            traceback.print_exc()
            self._send_response(500, {"success": False, "error": f"Internal server error: {e}"})
        finally:
             if scraper:
                 print("GDFLIX API: cloudscraper instance cleanup (automatic).")
                 pass


    def do_OPTIONS(self):
        self.send_response(204)
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Access-Control-Allow-Methods', 'POST, OPTIONS')
        self.send_header('Access-Control-Allow-Headers', 'Content-Type, x-token, User-Agent, Accept')
        self.send_header('Access-Control-Max-Age', '86400')
        self.end_headers()

    def _send_response(self, status_code, response_data):
        self.send_response(status_code)
        self.send_header('Content-type', 'application/json')
        self.send_header('Access-Control-Allow-Origin', '*')
        self.end_headers()
        self.wfile.write(json.dumps(response_data).encode('utf-8'))

File: api/gdflix.js
// Vercel Serverless Function for GDFLIX Bypass
// Adapted from the Node.js/Express version.

const axios = require('axios');
const cheerio = require('cheerio');

// --- Configuration ---
const HEADERS = {
    'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36',
    'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9',
    'Accept-Language': 'en-US,en;q=0.9',
    'Referer': 'https://google.com'
};
const GENERATION_TIMEOUT = 40 * 1000; // In milliseconds
const POLL_INTERVAL = 5 * 1000;      // In milliseconds
const REQUEST_TIMEOUT = 30 * 1000;   // In milliseconds
const MAX_REDIRECT_HOPS = 5;

// --- Utility Functions ---
const sleep = (ms) => new Promise(resolve => setTimeout(resolve, ms));

// --- Core GDFLIX Bypass Function (The "Engine") ---
async function getGdflixDownloadLink(startUrl) {
    const session = axios.create({
        headers: HEADERS,
        timeout: REQUEST_TIMEOUT,
        maxRedirects: 5
    });
    const logs = [];
    let currentUrl = startUrl;
    let hopsCount = 0;
    
    let page2_drivebot_url, html_content_p2_drivebot, page3_drivebot_url, html_content_p3_drivebot;

    try {
        while (hopsCount < MAX_REDIRECT_HOPS) {
            logs.push(`[Hop ${hopsCount}] Fetching/Checking URL: ${currentUrl}`);
            let response;
            try {
                response = await session.get(currentUrl, { validateStatus: () => true });
                if (response.status >= 400) {
                     logs.push(`  Error fetching ${currentUrl}: Status ${response.status}`);
                     return { finalLink: null, logs };
                }
            } catch (e) {
                logs.push(`  Error fetching ${currentUrl}: ${e.message}`);
                return { finalLink: null, logs };
            }
            const landedUrl = response.request.res.responseUrl || currentUrl;
            const htmlContent = response.data;
            logs.push(`  Landed on: ${landedUrl} (Status: ${response.status})`);
            let nextHopUrl = null;
            let isSecondaryRedirect = false;
            const metaMatch = htmlContent.match(/<meta\s+http-equiv="refresh"\s+content="[^"]*url=([^"]+)"/i);
            if (metaMatch) {
                const extractedUrl = metaMatch[1].trim().split(';')[0];
                const potentialNext = new URL(extractedUrl, landedUrl).href;
                if (potentialNext.split('#')[0] !== landedUrl.split('#')[0]) {
                    nextHopUrl = potentialNext;
                    logs.push(`  Detected META refresh redirect to: ${nextHopUrl}`);
                    isSecondaryRedirect = true;
                }
            }
            if (!isSecondaryRedirect) {
                const jsMatch = htmlContent.match(/location\.replace\(['"]([^'"]+)['"]/i);
                if (jsMatch) {
                    const extractedUrl = jsMatch[1].trim().split('+document.location.hash')[0].replace(/['"\s]/g, '');
                    const potentialNext = new URL(extractedUrl, landedUrl).href;
                    if (potentialNext.split('#')[0] !== landedUrl.split('#')[0]) {
                        nextHopUrl = potentialNext;
                        logs.push(`  Detected JS location.replace redirect to: ${nextHopUrl}`);
                        isSecondaryRedirect = true;
                    }
                }
            }
            if (isSecondaryRedirect && nextHopUrl) {
                logs.push(`  Following secondary redirect...`);
                currentUrl = nextHopUrl;
                hopsCount++;
                await sleep(500);
            } else {
                logs.push(`  No further actionable secondary redirect found. Proceeding with content analysis.`);
                currentUrl = landedUrl;
                logs.push(`--- Final Content Page HTML Snippet (URL: ${currentUrl}) ---`);
                logs.push(String(htmlContent).substring(0, 3000) + '...');
                logs.push(`--- End Final Content Page HTML Snippet ---`);
                const $1 = cheerio.load(htmlContent);
                const page1Url = currentUrl;
                const possibleTagsP1 = $1('a, button');
                logs.push(`Found ${possibleTagsP1.length} potential link/button tags on final content page (${page1Url}).`);
                
                // --- PRIORITY 1: Pixeldrain
                let pixeldrainLinkTag = null;
                possibleTagsP1.each((i, el) => {
                    if (/pixeldrain\s*(dl)?/i.test($1(el).text().trim())) {
                        pixeldrainLinkTag = el;
                        return false;
                    }
                });
                if (pixeldrainLinkTag) {
                    let pixeldrainHref = $1(pixeldrainLinkTag).attr('href') || $1(pixeldrainLinkTag).closest('form').attr('action');
                    if (pixeldrainHref) {
                        const pixeldrainFullUrl = new URL(pixeldrainHref, page1Url).href;
                        logs.push(`Success: Found Pixeldrain link URL: ${pixeldrainFullUrl}`);
                        return { finalLink: pixeldrainFullUrl, logs };
                    }
                }

                // --- PRIORITY 2: CLOUD DOWNLOAD [R2]
                let cloudR2LinkTag = null;
                possibleTagsP1.each((i, el) => {
                    if (/cloud\s+download\s+\[R2\]/i.test($1(el).text().trim())) {
                        cloudR2LinkTag = el;
                        return false;
                    }
                });
                 if (cloudR2LinkTag) {
                    let cloudR2Href = $1(cloudR2LinkTag).attr('href') || $1(cloudR2LinkTag).closest('form').attr('action');
                    if (cloudR2Href) {
                        const finalDownloadLink = new URL(cloudR2Href, page1Url).href;
                        logs.push(`Success: Found R2 download link: ${finalDownloadLink}`);
                        return { finalLink: finalDownloadLink, logs };
                    }
                }

                // --- PRIORITY 3: Fast Cloud Download
                let fastCloudLinkTag = null;
                possibleTagsP1.each((i, el) => {
                    if (/fast\s*cloud\s*(download|dl)/i.test($1(el).text().trim())) {
                        fastCloudLinkTag = el;
                        return false;
                    }
                });

                if (fastCloudLinkTag) {
                    let fastCloudHref = $1(fastCloudLinkTag).attr('href') || $1(fastCloudLinkTag).closest('form').attr('action');
                    if (fastCloudHref) {
                        const intermediateUrl = new URL(fastCloudHref, page1Url).href;
                        await sleep(1000);
                        const responseIntermediate = await session.get(intermediateUrl, { headers: { 'Referer': page1Url } });
                        const page2Url = responseIntermediate.request.res.responseUrl || intermediateUrl;
                        const $2 = cheerio.load(responseIntermediate.data);

                        let resumeLinkTag = null;
                        $2('a, button').each((i, el) => {
                            if (/cloud\s+resume\s+download/i.test($2(el).text().trim())) {
                                resumeLinkTag = el;
                                return false;
                            }
                        });
                        if (resumeLinkTag) {
                            let finalLinkHref = $2(resumeLinkTag).attr('href') || $2(resumeLinkTag).closest('form').attr('action');
                            if (finalLinkHref) {
                                const finalDownloadLink = new URL(finalLinkHref, page2Url).href;
                                return { finalLink: finalDownloadLink, logs };
                            }
                        }

                        let generateTag = $2('button#cloud').get(0);
                        if (!generateTag) {
                             $2('a, button').each((i, el) => {
                                if (/generate\s+cloud\s+link/i.test($2(el).text().trim())) {
                                    generateTag = el; return false;
                                }
                            });
                        }

                        if (generateTag) {
                            const postData = {};
                            $2(generateTag).closest('form').find('input[type="hidden"]').each((i, input) => {
                                const name = $2(input).attr('name');
                                if(name) postData[name] = $2(input).attr('value') || '';
                            });
                            const finalPostData = { ...{ action: 'cloud', key: '08df4425e31c4330a1a0a3cefc45c19e84d0a192', action_token: '' }, ...postData };
                            const postBody = new URLSearchParams(finalPostData).toString();
                            const postResponse = await session.post(page2Url, postBody, { headers: { 'Referer': page2Url, 'x-token': new URL(page2Url).hostname, 'X-Requested-With': 'XMLHttpRequest' } });
                            let responseData = postResponse.data;
                             if (typeof responseData !== 'object') { try { responseData = JSON.parse(responseData); } catch (e) { responseData = null; } }
                            
                            if (responseData && (responseData.visit_url || responseData.url)) {
                                const page3FcUrl = new URL(responseData.visit_url || responseData.url, page2Url).href;
                                const startTime = Date.now();
                                while (Date.now() - startTime < GENERATION_TIMEOUT) {
                                    await sleep(POLL_INTERVAL);
                                    const pollResponse = await session.get(page3FcUrl, { headers: { 'Referer': page3FcUrl } });
                                    const pollLandedUrl = pollResponse.request.res.responseUrl || page3FcUrl;
                                    const pollSoup = cheerio.load(pollResponse.data);
                                    let polledResumeTag = null;
                                    pollSoup('a, button').each((i, el) => {
                                        if (/cloud\s+resume\s+download/i.test(pollSoup(el).text().trim())) {
                                            polledResumeTag = el; return false;
                                        }
                                    });
                                    if (polledResumeTag) {
                                        let finalLinkHref = pollSoup(polledResumeTag).attr('href') || pollSoup(polledResumeTag).closest('form').attr('action');
                                        if (finalLinkHref) {
                                            return { finalLink: new URL(finalLinkHref, pollLandedUrl).href, logs };
                                        }
                                    }
                                }
                                logs.push('Error: Link generation timed out.');
                            }
                        }
                    }
                }

                // --- PRIORITY 4: Drivebot
                let drivebotInitialTag = null;
                possibleTagsP1.each((i, el) => { if(/DRIVEBOT/i.test($1(el).text().trim())) { drivebotInitialTag = el; return false; } });
                
                if (drivebotInitialTag) {
                    let drivebotInitialHref = $1(drivebotInitialTag).attr('href') || $1(drivebotInitialTag).closest('form').attr('action');
                    if(drivebotInitialHref){
                        const drivebotStep1Url = new URL(drivebotInitialHref, page1Url).href;
                        const res_db_s1 = await session.get(drivebotStep1Url, { headers: { 'Referer': page1Url } });
                        page2_drivebot_url = res_db_s1.request.res.responseUrl || drivebotStep1Url;
                        const $d2 = cheerio.load(res_db_s1.data);
                        let serverChoiceTag = null;
                        $d2('a, button').each((i, el) => { if(/DRIVEBOT\s*1/i.test($d2(el).text().trim())){ serverChoiceTag = el; return false; } });
                        if(!serverChoiceTag) $d2('a, button').each((i, el) => { if(/DRIVEBOT/i.test($d2(el).text().trim())){ serverChoiceTag = el; return false; } });

                        if(serverChoiceTag){
                            let nextUrl, payload = {}, method = 'GET';
                            const $serverTag = $d2(serverChoiceTag);
                            const parentForm = $serverTag.closest('form');
                            if(parentForm.length > 0){
                                nextUrl = new URL(parentForm.attr('action') || '', page2_drivebot_url).href;
                                method = (parentForm.attr('method') || 'GET').toUpperCase();
                                parentForm.find('input').each((i, input) => {
                                    const name = $d2(input).attr('name');
                                    if(name) payload[name] = $d2(input).attr('value') || '';
                                });
                            } else if (serverChoiceTag.name === 'a') {
                                nextUrl = new URL($serverTag.attr('href'), page2_drivebot_url).href;
                            }

                            if(nextUrl){
                                const res_db_s2 = method === 'POST' ? await session.post(nextUrl, new URLSearchParams(payload).toString(), { headers: { 'Referer': page2_drivebot_url } }) : await session.get(nextUrl, { params: payload, headers: { 'Referer': page2_drivebot_url } });
                                page3_drivebot_url = res_db_s2.request.res.responseUrl || nextUrl;
                                const $d3 = cheerio.load(res_db_s2.data);
                                let genLinkButton = null;
                                $d3('a, button, input').each((i, el) => { if(/Generate Link/i.test($d3(el).text().trim() || $d3(el).attr('value'))) { genLinkButton = el; return false; } });

                                if(genLinkButton){
                                    let postUrlGen = page3_drivebot_url, postDataGen = {}, httpMethodGen = 'POST';
                                    const parentFormGen = $d3(genLinkButton).closest('form');
                                    if(parentFormGen.length > 0) {
                                        postUrlGen = new URL(parentFormGen.attr('action') || '', page3_drivebot_url).href;
                                        httpMethodGen = (parentFormGen.attr('method') || 'POST').toUpperCase();
                                        parentFormGen.find('input').each((i, input) => {
                                            const name = $d3(input).attr('name');
                                            if(name) postDataGen[name] = $d3(input).attr('value') || '';
                                        });
                                    }
                                    const resGen = httpMethodGen === 'POST' ? await session.post(postUrlGen, new URLSearchParams(postDataGen).toString(), { headers: { 'Referer': page3_drivebot_url } }) : await session.get(postUrlGen, { params: postDataGen, headers: { 'Referer': page3_drivebot_url } });
                                    const $d4 = cheerio.load(resGen.data);
                                    let finalDlLink = null;
                                    const linkPattern = /https?:\/\/[^\s"'`]*\.gdindex\.lol[^\s"'`]*/;
                                    const linkInputTag = $d4('input').filter((i, el) => linkPattern.test($d4(el).attr('value')));
                                    if(linkInputTag.length > 0) finalDlLink = linkInputTag.attr('value').trim();
                                    if(!finalDlLink){
                                        const linkAnchorTag = $d4('a').filter((i, el) => linkPattern.test($d4(el).attr('href')));
                                        if (linkAnchorTag.length > 0) finalDlLink = linkAnchorTag.attr('href').trim();
                                    }
                                    if (finalDlLink) return { finalLink: finalDlLink, logs };
                                }
                            }
                        }
                    }
                }

                logs.push("Error: All prioritized search attempts failed to yield a download link.");
                return { finalLink: null, logs };
            }
        }
        if (hopsCount >= MAX_REDIRECT_HOPS) {
            logs.push(`Error: Exceeded maximum redirect hops (${MAX_REDIRECT_HOPS}).`);
        }
    } catch (e) {
        logs.push(`FATAL: An unexpected error occurred: ${e.message}\n${e.stack}`);
        return { finalLink: null, logs };
    }
    return { finalLink: null, logs };
}

// --- Vercel Serverless Function Handler ---
export default async function handler(request, response) {
    // Set CORS headers
    response.setHeader('Access-Control-Allow-Origin', '*'); // Or a specific domain
    response.setHeader('Access-Control-Allow-Methods', 'POST, OPTIONS');
    response.setHeader('Access-Control-Allow-Headers', 'Content-Type');

    if (request.method === 'OPTIONS') {
        return response.status(200).end();
    }

    if (request.method !== 'POST') {
        response.setHeader('Allow', ['POST']);
        return response.status(405).json({ error: 'Method Not Allowed' });
    }

    let scriptLogs = [];
    const result = { success: false, error: "Request processing failed", finalUrl: null, logs: scriptLogs };
    let statusCode = 500;

    try {
        const { gdflixUrl } = request.body;
        if (!gdflixUrl) {
            result.error = "Invalid or missing JSON (expected {'gdflixUrl': '...' })";
            return response.status(400).json(result);
        }
        scriptLogs.push(`Received gdflixUrl: ${gdflixUrl}`);

        const { finalLink, logs: scriptLogsFromFunc } = await getGdflixDownloadLink(gdflixUrl);
        scriptLogs.push(...scriptLogsFromFunc);

        if (finalLink) {
            scriptLogs.push("Bypass process completed successfully.");
            result.success = true;
            result.finalUrl = finalLink;
            result.error = null;
            statusCode = 200;
        } else {
            scriptLogs.push("Bypass process failed to find the final download link.");
            result.success = false;
            let extractedError = "GDFLIX Extraction Failed (Check logs for details)";
            // Simple error extraction for now
             const lastErrorLog = scriptLogs.filter(log => /Error:/i.test(log)).pop() || '';
             if (lastErrorLog) {
                 extractedError = lastErrorLog.split(/Error:\s*/i)[1] || lastErrorLog;
             }
            result.error = extractedError.substring(0, 250);
            statusCode = 200; // Return 200 on "user error" like a failed bypass
        }
    } catch (e) {
        console.error(`FATAL API Handler Error: ${e.message}`, e.stack);
        scriptLogs.push(`FATAL API Handler Error: An unexpected server error occurred.`);
        result.error = "Internal server error processing the request.";
        statusCode = 500;
    } finally {
        result.logs = scriptLogs;
        response.status(statusCode).json(result);
    }
}
